tools/rtpsrelay/RelayHandler.cpp

```diff
@@ -431,6 +345,154 @@ bool VerticalHandler::parse_message(OpenDDS::RTPS::MessageParser& message_parser
   return true;
 }
 
+void VerticalHandler::send(const RelayAddressesMap& relay_addresses_map,
+                           const OpenDDS::DCPS::Message_Block_Shared_Ptr& msg)
+{
+  size_t fan_out = 0;
+  for (const auto& p : relay_addresses_map) {
+    const auto& addrs = p.first;
+    const auto& guids = p.second;
+    fan_out += guids.size();
+    if (addrs != relay_addresses_) {
+      horizontal_handler_->enqueue_message(extract_relay_address(addrs), guids, msg);
+    } else {
+      for (const auto& guid : guids) {
+        const auto p = find(guid);
+        if (p != end()) {
+          for (const auto& addr : p->second) {
+            enqueue_message(addr, msg);
+          }
+        } else {
+          ACE_ERROR((LM_WARNING, "(%P|%t) %N:%l WARNING: VerticalHandler::send failed to get address\n"));
+        }
+      }
+    }
+  }
+  max_fan_out(fan_out);
+}
+
+HorizontalHandler::HorizontalHandler(ACE_Reactor* reactor,
+                                     const AssociationTable& association_table)
+  : RelayHandler(reactor, association_table)
+  , vertical_handler_(nullptr)
+{}
+
+void HorizontalHandler::enqueue_message(const std::string& addr,
+                                        const GuidSet& guids,
+                                        const OpenDDS::DCPS::Message_Block_Shared_Ptr& msg)
+{
+  // Determine how many guids we can pack into a single UDP message.
+  const auto max_guids_per_message = (OpenDDS::DCPS::TransportSendStrategy::UDP_MAX_MESSAGE_SIZE - msg->length() - 4) / sizeof(OpenDDS::DCPS::RepoId);
+
+  auto remaining = guids.size();
+  auto pos = guids.begin();
+
+  while (remaining) {
+    auto guids_in_message = std::min(max_guids_per_message, remaining);
+    remaining -= guids_in_message;
+
+    RelayHeader relay_header;
+    auto& to = relay_header.to();
+    for (; guids_in_message; --guids_in_message, ++pos) {
+      to.push_back(repoid_to_guid(*pos));
+    }
+
+    size_t size = 0, padding = 0;
+    OpenDDS::DCPS::gen_find_size(relay_header, size, padding);
+    ACE_Message_Block* header_block = new ACE_Message_Block(size + padding);
+    OpenDDS::DCPS::Serializer ser(header_block);
+    ser << relay_header;
+    header_block->cont(msg.get()->duplicate());
+    RelayHandler::enqueue_message(addr, OpenDDS::DCPS::Message_Block_Shared_Ptr(header_block));
```

Comment:
Since `header_block`  will be managed by Shared_Ptr, it would be cleaner to start it out that way when it's allocated on line 402.